utils/waitForValidatorSetChange.js

```diff
@@ -0,0 +1,32 @@
+const getContract = require('./getContract');
+
+const RETRY_INTERVAL_MS = 2499;
+
+module.exports = async (web3) => {
+    let ValidatorSetAuRaContract = getContract('ValidatorSetAuRa', web3).instance;
+    let StakingAuRaContract = getContract('StakingAuRa', web3).instance;
+
+    let initialStakingEpoch = parseInt(await StakingAuRaContract.methods.stakingEpoch().call());
+    let epochChangeBlock;
+
+    // wait for the next staking epoch
+    while (true) {
+        await new Promise(r => setTimeout(r, RETRY_INTERVAL_MS));
+        let currentStakingEpoch = parseInt(await StakingAuRaContract.methods.stakingEpoch().call());
+        let currentBlock = parseInt((await web3.eth.getBlock('latest')).number);
+        if (currentStakingEpoch > initialStakingEpoch) {
+            console.log(`***** Staking epoch changed at block ${currentBlock} (new epoch ${currentStakingEpoch})`);
+            epochChangeBlock = currentBlock;
+            break;
+        }
+    }
+
+    // wait a few more blocks to finalize the change
+    while (true) {
+        await new Promise(r => setTimeout(r, RETRY_INTERVAL_MS));
+        let currentBlock = parseInt((await web3.eth.getBlock('latest')).number);
+        if (currentBlock - epochChangeBlock >= 6) {
```

Comment:
Let's use here [`validatorSetApplyBlock()`](https://github.com/poanetwork/posdao-contracts/blob/5326d9eaafac4cfc113992ba0cfe66d8bab41ca7/contracts/abstracts/ValidatorSetBase.sol#L268) getter instead: when the staking epoch number is increased, the [`validatorSetApplyBlock` becomes `0`](https://github.com/poanetwork/posdao-contracts/blob/5326d9eaafac4cfc113992ba0cfe66d8bab41ca7/contracts/abstracts/ValidatorSetBase.sol#L441-L445) until the new validator set [is applied by Parity](https://github.com/poanetwork/posdao-contracts/blob/5326d9eaafac4cfc113992ba0cfe66d8bab41ca7/contracts/abstracts/ValidatorSetBase.sol#L84).

So, we could call `validatorSetApplyBlock()` inside this `while` and check if it returns `0`. If it does, then `continue` the loop, otherwise `break`. Thus, we could remove `currentBlock` and `epochChangeBlock` variables inside this `while`.

Reply:
This is a very cool function.
I also made `waitForValidatorSetChange` return the new validator set and further decreased staking epoch duration to 40 blocks, otherwise tests were taking about 8 minutes.